game.js

```diff
@@ -750,76 +732,90 @@ module.exports = function createGame(options) {
             }
             // Assign the roles the player selected.
-            for (i = 0; i < playerState.influence.length; i++) {
-                if (!playerState.influence[i].revealed) {
-                    playerState.influence[i].role = command.roles.pop()
+            for (var i = 0; i < state.players[playerIdx].influence.length; i++) {
+                if (!state.players[playerIdx].influence[i].revealed) {
+                    state.players[playerIdx].influence[i].role = command.roles.pop()
                 }
             }
             // Return the other roles to the deck.
             deck = shuffle(deck.concat(unchosen));
             addHistory('exchange', curTurnHistGroup(), '{%d} exchanged roles', playerIdx);
             nextTurn();
-
-        } else if (command.command == 'interrogate') {
-            if (state.state.name != stateNames.INTERROGATE) {
-                throw new GameException('Incorrect state');
-            }
-            if (state.state.playerIdx != playerIdx) {
-                throw new GameException('Not your turn');
-            }
+        }
+    ];
+    behaviors[['interrogate', stateNames.INTERROGATE]] = [
+        playersTurn,
+        (command, playerIdx) => {
             // Send a history event only to the player who was interrogated.
             addHistoryAsync(
                 state.state.target,
                 'interrogate',
                 curTurnHistGroup(),
                 format('{%d} saw your %s', playerIdx, state.state.confession));
             if (command.forceExchange) {
-                var target = state.players[state.state.target];
-                var idx = indexOfInfluence(target, state.state.confession);
+                const target = state.players[state.state.target];
+                const idx = indexOfInfluence(target, state.state.confession);
                 if (idx == null) {
                     throw new GameException('Target does not have the confessed role');
                 }
                 target.influence[idx].role = swapRole(state.state.confession);
                 addHistory('interrogate', curTurnHistGroup(), '{%d} forced {%d} to exchange roles', playerIdx, state.state.target);
+                nextTurn();
+                return;
             }
-            else {
-                addHistory('interrogate', curTurnHistGroup(), '{%d} allowed {%d} to keep the same roles', playerIdx, state.state.target);
-            }
+            addHistory('interrogate', curTurnHistGroup(), '{%d} allowed {%d} to keep the same roles', playerIdx, state.state.target);
             nextTurn();
+        }
+    ];
 
-        } else {
-            throw new GameException('Unknown command');
+    function command(playerIdx, command) {
+        debug('command from player: ' + playerIdx);
+        debug(command);
+        if (!state.players[playerIdx]) {
+            throw new GameException('Unknown player');
+        }
+        if (command.stateId != state.stateId) {
```

Comment:
The original file had a comment `You can always leave, even if your state id is old` - this is important